adr/ADR-42.md

```diff
@@ -0,0 +1,168 @@
+
+**NOTE: We should not describe this in terms of exclusivity as there is no such guarantee, there will be times when one
+client think it is pinned when it is not because the server switched.** 
+
+If the current pinned client is priority 5 and a pull comes in for this group from a lower priority client the server will
+switch to this lower priority client.
+
+The `pinned_client` policy provides server-side orchestration for the selection of the pinned client.
+
+```go
+{
+	PriorityGroups: ["jobs"],
+	PriorityPolicy: "pinned_client",
+	PriorityTimeout: 120*time.Second,
+	AckPolicy: "explicit",
+	// ... other consumer options
+}
+```
+
+This configuration states:
+
+ * We have 1 group defined and all pulls have to belong to this group
+ * The policy is `pinned_client` that activates these behaviors
+ * When a pinned client has not done any pulls in the last 120 seconds the server will switch to another client
+ * AckPolicy has to be `explicit`. If we force this ack policy in normal use we should error in Pedantic mode
+
+A pull request will have the following additional fields:
+
+ * `"group": "jobs"` - the group the pull belongs to, pulls not part of a valid group will result in an error
+ * `"priority": 1` - the priority this pull is done at, valid values are 1 to 10. Absent means 5. Anything else is an error
+ * `"id": "xyz"` - the pinned client will have this ID set to the one the server supplied (see below), otherwise this field is absent
+
+When no pinned client has been selected by the server the first message that will be delivered will include a `Nats-Pin: xyz`
+header. The client that gets this message should at that point ensure that all future pull requests have the same ID set.
+The pinned client is chosen by walking all the waiting pulls and picking the lowest priority one.
+
+If a lower priority pull is received than the current pinned clients most recent pull the server will switch to a new
+pinned client:
+
+ 1. Stop delivering messages for this group, wait for all in-flight messages to be completed, continue to serve heartbeats etc 
+ 2. Pick the new pinned client by priority and group
+ 3. Store the new pinned `nuid`
+ 4. Deliver the message to the new pinned client with the ID set
+ 5. Create an advisory that a new pinned client was picked
+ 6. Respond with a 4xx header to any pulls, including waiting ones, that have a different ID set. Client that received this error will clear the ID and pull with no ID 
+
+If no pulls from the pinned client is received within `PriorityTimeout` the server will switch again using the same flow as above.
+
+Clients can expose call-back notifications when they become pinned (first message with `Nats-Pin` header is received) and
+when they lose the pin (they receive the 4xx error when doing a pull with a old ID).
+
+A new API, `$JS.API.CONSUMER.UNPIN.<STREAM>.<CONSUMER>.<GROUP>`, can be called which will clear the ID and trigger a client switch as above.
```

Comment:
`SELECT` - pin is weird to me

Reply:
its definitely not selecting, its more clearing the pin and then later doing a switch.

We had a lot of to and fro on the name and `pin` really worked for most people :)